src/main/java/com/shapesecurity/shift/es2016/semantics/visitor/Truthiness.java

```diff
@@ -0,0 +1,316 @@
+        return Maybe.of(true);
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(LiteralFunction literalFunction) {
+        return Maybe.of(true);
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(LiteralInfinity literalInfinity) {
+        return Maybe.of(true);
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(LiteralNull literalNull) {
+        return Maybe.of(false);
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(LiteralNumber literalNumber) {
+        return Maybe.of(literalNumber.value != 0 && !Double.isNaN(literalNumber.value));
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(LiteralRegExp literalRegExp) {
+        return Maybe.of(true);
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(LiteralString literalString) {
+        return Maybe.of(!(literalString.value.isEmpty()));
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(LiteralSymbol literalSymbol) {
+        return Maybe.empty();
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(LiteralUndefined literalUndefined) {
+        return Maybe.of(false);
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(MemberAccess memberAccess) {
+        return Maybe.empty();
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(MemberAssignment memberAssignment) {
+        return Maybe.empty();
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(New new_) {
+        return Maybe.of(true);
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(RequireObjectCoercible requireObjectCoercible) {
+        return Maybe.empty();
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(LocalReference localReference) {
+        return Maybe.empty();
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(TemporaryReference temporaryReference) {
+        return Maybe.empty();
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(This this_) {
+        return Maybe.empty();
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(TypeCoercionNumber typeCoercionNumber) {
+        return Maybe.empty();
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(TypeCoercionString typeCoercionString) {
+        return Maybe.empty();
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(TypeofGlobal typeofGlobal) {
+        return Maybe.of(true);
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(VariableAssignment variableAssignment) {
+        return Maybe.empty();
+    }
+
+    @Nonnull
+    @Override
+    public Maybe<Boolean> apply(Equality equality) {
+        NodeWithValue left = equality.left();
+        NodeWithValue right = equality.right();
+        if (ConstantFolder.isPure(left).orJust(false) && left.equals(right)) {
```

Comment:
`isPure` isn't correct here. We should instead use `ToPrimitive` like this:

```java
if (ToPrimitive(left).flatMap(l -> ToPrimitive(right).map(r -> l.equals(r))).orJust(false)) {
```

Alternatively, we could just do

```java
if (Type(left).map(t -> t != Type.Object).orJust(false) && left.equals(right)) {
```

That's probably simpler.